dask/dataframe/core.py

```diff
@@ -4204,17 +4204,28 @@ def _take_last(a, skipna=True):
         Whether to exclude NaN
 
     """
+    def _last_valid(s):
+        for i in range(1, min(10, len(s) + 1)):
+            val = s.iloc[-i]
+            if not pd.isnull(val):
+                return val
+        else:
+            nonnull = s[s.notna()]
+            if not nonnull.empty:
+                return nonnull.iloc[-1]
+        return None
+
     if skipna is False:
         return a.iloc[-1]
     else:
         # take last valid value excluding NaN, NaN location may be different
-        # in each columns
-        group_dummy = np.ones(len(a.index))
-        last_row = a.groupby(group_dummy).last()
-        if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-            return pd.Series(last_row.values[0], index=a.columns)
+        # in each column
+        if is_dataframe_like(a):
+            # create Series of appropriate backend dataframe library
+            series_typ = type(a.iloc[0])
```

Comment:
Hrm, this will fail if there are empty partitions, as in this example.

```python
import pandas as pd
import dask.dataframe as dd
df = pd.DataFrame({'x': [1, 2, 3, 4, 5, 6, 7, 8]})
ddf = dd.from_pandas(df, npartitions=4)
ddf[ddf.x < 5].cumsum().compute()
```

(sorry for not noticing this before)

Maybe `type(a.head(0).sum())` ?  Actually no, cudf doesn't support this.  We'll need to find some cheap way to get the Series type of a DataFrame object that always works :/

Reply:
Good catch. Will think about this.

Reply:
Could we leverage the fact that slices behave differently? Instead of grabbing with `a.iloc[0]` let's perhaps grab with `a.iloc[:0]` or `a.iloc[0:1]`? This appears to solve the indexing issue but results in a downstream issue still. Looking into it.

Reply:
Also, as a note, it looks like the original `_take_last` also fails when there are empty partitions.

```python
...
/Users/nickbecker/NVIDIA/dask/dask/dataframe/core.py in _take_last(a, skipna)
   4213         last_row = a.groupby(group_dummy).last()
   4214         if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-> 4215             return pd.Series(last_row.values[0], index=a.columns)
   4216         else:
   4217             return last_row.values[0]

IndexError: index 0 is out of bounds for axis 0 with size 0
```

Reply:
What if it has no columns?

Reply:
If this failed before then we can probably leave this as is and raise an issue.

Reply:
Ah, seems like you just did this in https://github.com/dask/dask/issues/4743

Reply:
Yep. Defer to you on whether you want to leave this as is since it was already failing on empty partition dataframes or try to fix it in this PR.


The above snippet does solve the problem for cumsum and cumprod, but still causes issues with cummin and cummax when it tries to compare a non-empty series with an empty series:

```python
/dask/dask/dataframe/methods.py in cummin_aggregate(x, y)
    147 def cummin_aggregate(x, y):
    148     if isinstance(x, (pd.Series, pd.DataFrame)):
--> 149         return x.where((x < y) | x.isnull(), y, axis=x.ndim - 1)
    150     else:       # scalar
    151         return x if x < y else y

/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/ops.py in wrapper(self, other, axis)
   1674
   1675         elif isinstance(other, ABCSeries) and not self._indexed_same(other):
-> 1676             raise ValueError("Can only compare identically-labeled "
   1677                              "Series objects")
   1678

ValueError: Can only compare identically-labeled Series objects
ipdb> self
x    0
dtype: int64
ipdb> other
Series([], dtype: float64)
ipdb>
```

Reply:
I'd be happy if you wanted to include it.  You might also include a test with the currently passing cum-reductions (and, if you wanted to get fancy, xfailed parameters for the failing ones).

```python
@pytest.mark.parametrize('func', [
    M.cumsum,
    M.cumprod,
    pytest.param(M.cummin, marks=[pytest.mark.xfail(reason="...")]),
    ...
])
def test_cum_reduction_empty_partitions(func):
    df = ...
    ...
```

(I could be totally wrong about the `pytest.param` syntax by the way.  I would check existing tests.

Reply:
I'll include the snippet and add tests for the empty partition case so that we still can at least pass cumsum/cumprod with empty partitions, which is an improvement on the current situation. I'll match the xfail style for the cummin/cummax on empty partitions to the style in the other tests.